src/core/lib/resolver/server_address.h

```diff
@@ -32,44 +31,74 @@
 
 namespace grpc_core {
 
-//
-// ServerAddress
-//
-
-// A server address is a grpc_resolved_address with an associated set of
-// channel args.  Any args present here will be merged into the channel
-// args when a subchannel is created for this address.
-class ServerAddress {
+class ResolverAttributeMap {
  public:
   // Base class for resolver-supplied attributes.
-  // Unlike channel args, these attributes don't affect subchannel
-  // uniqueness or behavior.  They are for use by LB policies only.
-  //
-  // Attributes are keyed by a C string that is unique by address, not
-  // by value.  All attributes added with the same key must be of the
-  // same type.
   class AttributeInterface {
    public:
     virtual ~AttributeInterface() = default;
 
+    // The type name for this attribute.
+    // There can be only one attribute of a given type in a given AttributeMap.
+    // Attributes are keyed by the address of the string, not the value.
```

Comment:
Is there an incredibly strong justification for this?

Reply:
Providing a string instance does not solve this problem, and the design doesn't work.

C++ implementations are allowed to (but not required to) merge duplicate string values.
You've recently seen cases where the same string token was given two pointers in one build system.
It would be a valid compile for two instances of `"foo"` in the same binary, coming from two different libraries, to be given the same pointer also.
C++ does not provide the guarantees this code is asking for.

Suggestions to improve the design:
1. mandate a namespace prefix (`grpc.lb_attr1`, `grpc.lb_attr2`, `customer.lb_attr1`) as we do for channel args (indeed the reason we do this for channel args)
2. implement an attribute class that actually provides the guarantees you're looking for (probably internally calling new on some non-zero length data structure to ensure a unique pointer), use that throughout instead of `const char *`.

Reply:
Thanks for pointing this out!

I've sent #29709 to introduce a `UniqueTypeName` API along the lines of your second suggestion and to change the other places in our code where we're using this pattern to use `UniqueTypeName` instead.

I've now merged that PR into this one and changed the resolver attribute API to also use `UniqueTypeName`.